src/Qir/Tests/QIR-dynamic/qsharp/qir-test-assert.qs

```diff
@@ -1,15 +1,363 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+// https://docs.microsoft.com/en-us/azure/quantum/concepts-the-qubit#single-qubit-operations    Qubit and Operations
+// https://docs.microsoft.com/en-us/azure/quantum/user-guide/libraries/standard/prelude         Operation Details, Multi-Qubit Operations
+
 namespace Microsoft.Quantum.Testing.QIR  {
 
     open Microsoft.Quantum.Diagnostics;
+    open Microsoft.Quantum.Intrinsic;
+
+    // To avoid the potentially circular dependency on QuantumLibraries repo we redefine the following {
+    // Identifiers from Microsoft.Quantum.Arrays:
+    // QuantumLibraries\Standard\src\Arrays\Arrays.qs:
+    function Head<'A> (array : 'A[]) : 'A {
+        //EqualityFactB(Length(array) > 0, true, $"Array must be of the length at least 1");
+        if Length(array) == 0 {
+            fail "Array must be of the length at least 1";
+        }
+        return array[0];
+    }
+    function Rest<'T> (array : 'T[]) : 'T[] {
+        return array[1 .. Length(array) - 1];
+    }
+    function ConstantArray<'T> (length : Int, value : 'T) : 'T[] {
+        mutable arr = new 'T[length];
+
+        for i in 0 .. length - 1 {
+            set arr w/= i <- value;
+        }
+
+        return arr;
+    }
+    function Most<'T> (array : 'T[]) : 'T[] {
+        return array[0 .. Length(array) - 2];
+    }
+
+    // QuantumLibraries\Standard\src\Arrays\Zip.qs:
+    function Zipped<'T, 'U>(left : 'T[], right : 'U[]) : ('T, 'U)[] {
+        let nElements = Length(left) < Length(right)
+                        ? Length(left)
+                        | Length(right);
+        mutable output = new ('T, 'U)[nElements];
+
+        for idxElement in 0 .. nElements - 1 {
+            set output w/= idxElement <- (left[idxElement], right[idxElement]);
+        }
+
+        return output;
+    }
+
+    // qsharp-runtime\src\Simulation\QSharpFoundation\Arrays\Enumeration.qs:
+    function IndexRange<'TElement>(array : 'TElement[]) : Range {
+       return 0..(Length(array) - 1);
+    }
+
+    // Identifiers from open Microsoft.Quantum.Canon:
+    // QuantumLibraries\Standard\src\Canon\Combinators\ApplyToEach.qs:
+    operation ApplyToEachCA<'T> (singleElementOperation : ('T => Unit is Adj + Ctl), register : 'T[])
+    : Unit is Adj + Ctl {
+        for idxQubit in IndexRange(register) {
+            singleElementOperation(register[idxQubit]);
+        }
```

Comment:
For now I'd prefer to keep my redefinitions as close as possible to the original definitions (as was explained in a different discussion earlier in this PR). To avoid adding bugs, or to cause suspicions if anyone later needs to investigate the bugs.